### ISSUE TITLE
Emitting bad code with `when (false)`

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#18678]
```C#

        [Fact]
        public void PEVerifyErrorWithFalse()
        {
            string source = @"
class C
{
    static void Main()
    {
        try
        {
            throw new System.Exception();
        }
        catch (System.Exception) when (false)
        {
            System.Console.Write(""catch"");
        }
    }
}
";
            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.Regular7_1, options: TestOptions.DebugExe);
            comp.VerifyDiagnostics(
                // (10,40): warning CS7095: Filter expression is a constant, consider removing the filter
                //         catch (System.Exception) when (false)
                Diagnostic(ErrorCode.WRN_FilterIsConstant, "false").WithLocation(10, 40),
                // (12,13): warning CS0162: Unreachable code detected
                //             System.Console.Write("catch");
                Diagnostic(ErrorCode.WRN_UnreachableCode, "System").WithLocation(12, 13)
                );
            //CompileAndVerify(comp); // PEVerify failed with Branch out of the method. Follow-up issue: https://github.com/dotnet/roslyn/issues/18609
        }
```

**Comment by elijah6** [comment]
Although this isn't marked as up-for-grabs, I had a look as it was related to #18750 , which I recently worked on. If it was already analysed and the solution is already being prepared, please ignore...

**Problem summary**
All catch clauses, as well as the end of the try-block, are unreachable, so the implicit return statement is not inserted at the end. But each unreachable catch block emits branching code (`leave.s ADDR`) to the end of the method expecting the implicit return, and this makes the code invalid.
```
      // ... Handler code...

      // [31 9 - 31 10]
      IL_002a: nop          
      IL_002b: leave.s      IL_002d
    } // end of filter handler

  } // end of method Program::Method2
```

**Options**
*option 1. Add implicit return instruction (`ret`) to satisfy the `leave.s` opcode that branches out from the end of the catch body.*
This will be a dummy ret instruction that will never be used because the end of the method body is unreachable. I thought about analysing both 'True' and 'False' cases at [VisitCatchBlock](https://github.com/dotnet/roslyn/blob/614299ff83da9959fa07131c6d0ffbc58873b6ae/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs#L104) similarly to If-statement, but that will make the flow analysis to behave as if the body of the filter expression is reachable. Since [Analyze(...)](https://github.com/dotnet/roslyn/blob/d0d07e5880947837f88af051fc1408ea952fcc98/src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs#L41) should really return `false` as the end of the method block is unreachable by definition, we would need another condition to make sure the implicit return statement is appended while `Rewrite()` is happening. 

Things to note;
- We are adding a dummy return just to satisfy `leave.s` that will never get executed (may not be a problem at all)
- Debugger can manually step into the 'unreachable' block (e.g. via VisualStudio, manually dragging the breakpoint into the block). Normally it should be fine as it will eventually leave the catch block, but what if the method has a return type and the catch block didn't have any? 
e.g. 
```C#
    private static bool MethodExample()
    {
        try
        {
            throw new Exception();
        }
        catch (Exception) when (false)
        {
            Console.WriteLine("Test2");
        }
    } 
```
This will be fine because the control flow analysis will think the end of the block is unreachable, and the only way out from the method is the `throw` statement in the try-block. But if you move the debugger step into the catch block manually, the dummy `ret` can return some random thing in the local stack. 


*option 2. Remove the catch handler block*
If-statement apparently does not emit the block if the condition if constant `false`. Maybe we can do a similar thing for the false filter expression [(around here in EmitStatement.cs)](https://github.com/dotnet/roslyn/blob/701ca5b2ce7e331d3bb47e55d842a8a1634e501d/src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs#L1045)? Simply replacing everything with `nop` won't work as the block would expect some leave statement, and adding `endfinally` would also complain as it is not in the `finally` block. So we would need some alternative. 

**Additional question**
Whichever method we choose, do we need to emit the unreachable block? Should we just follow what if-statement does and not emit the block at all?

**Comment by jcouv** [comment]
Thanks @elijah6 for the investigation. I don't think anyone drilled into this issue yet.
Adding @gafter @VSadov

**Comment by gafter** [comment]
Another option might be to drop the entire catch clause.

**Comment by tmat** [comment]
We need to emit all unreachable code in Debug builds as it can be reached by modifying the method or manually setting the IP.

**Comment by tmat** [comment]
Related: https://github.com/dotnet/roslyn/issues/17947

**Comment by VSadov** [comment]
The simplest solution would be to make the unreachable leaving branch to leave to itself. - unreachable infinite loop similar to what is done for leaves that are blocked by finallies that never terminate.

In addition it would be nice to remove these catches entirely. 
However although in some cases catch may become unreachable as a result of other optimizations so the code that handles such catches in the codegen is still needed -in case not all are optimized away earlier.

**Comment by VSadov** [comment]
@tmat  - keeping unreachable code could be tricky. It can violate numerous IL invariants - unbalanced stacks, reading uninitialized data, missing returns,dropping through the end of EH regions, branching to nowhere... 

C# compiler was always removing unreachable code. It is much easier to detect and drop it than ensure that it is well formed. 

Not that it is impossible, just to mention that it is not a trivial thing to do.

**Comment by tmat** [comment]
@vsadov I think we need to distinguish unreachable code within statement, e.g. ```F(throw null, Unreachable())``` vs. unreachable statements (or more generally sequence points). The former is fine to elide. The latter should be preserved. In the latter case I don't think we should get into unbalanced stack issues, since statements always start with empty stack. Not sure about the other issues you mention, can you give examples?

**Comment by gafter** [comment]
We do not preserve unreachable statements. The following hypothetical unit test passes.

``` c#
        [Fact, WorkItem(18678, "https://github.com/dotnet/roslyn/issues/18678")]
        public void UnreachableCodeEliminated()
        {
            var source =
@"using System;
class Program
{
    public static void Main(string[] args)
    {
    }
    public static int M1(int i) => i + 1;
    public static int M(int i)
    {
        throw null;
        Console.Write(M1(2));
    }
}";
            var compVerifier = CompileAndVerify(source,
                options: TestOptions.DebugDll.WithOutputKind(OutputKind.ConsoleApplication),
                parseOptions: TestOptions.Regular7_1,
                expectedOutput: "");
            compVerifier.VerifyIL("Program.M",
@"{
  // Code size        3 (0x3)
  .maxstack  1
  IL_0000:  nop
  IL_0001:  ldnull
  IL_0002:  throw
}"
            );
        }
```

**Comment by tmat** [comment]
@gafter I guess that's ok. I was suggesting we might want to to improve debugging in some cases but it's not that important.

**Comment by tmat** [comment]
@gafter @VSadov Should local variables and constants be emitted when defined in unreachable code?

```C#
class C
{
    void F()
    {
        int a = 1;
        throw null;
        { int b = 1; }
    }
}
```

Emits 
```
{
  // Code size        5 (0x5)
  .maxstack  1
  .locals init (int V_0, //a
                int V_1) //b
  IL_0000:  nop
  IL_0001:  ldc.i4.1
  IL_0002:  stloc.0
  IL_0003:  ldnull
  IL_0004:  throw
}
```

**Comment by VSadov** [comment]
@tmat - you should enter another bug for keeping unreachable code in debug emit, where we can discuss these things.

Right now, and as it always have been - dead code elimination is not just an optimization, since it is not optional, and we do it in debug as well. - It is needed to ensure that IL is well-formed. There could be other ways, but this is how it is done now.

For the unused locals - Unused locals do not make IL unverifiable or otherwise bad, so we do not have to remove them.

**Comment by VSadov** [comment]
The final dead code elimination pass is performed when the code is in the form of IL Basic Blocks. It is basically just the control flow graph for the IL we are about to emit and we mark-and-sweep closures of the graph that are not reachable form the entry points. 
There is not a lot of distinction between statements and expressions at that point.

What you are suggesting would be better handled by "patching up" the bound tree in debug to ensure it is all emittable and just suppress the DCE in codegen.

It is a bigger issue than what we have here.

**Comment by elijah6** [comment]
Based on the discussion so far, I have made a change to ignore the entire catch clause if the filter expression is false as @gafter suggested.

**Comment by gafter** [comment]
Fixed in #19294
